Continue porting of graphics related types to the new IPC serialization format

gavin-apple opened the pull request:
#### 4f6caeb7a6616e3ab6f4ea54a9ce483ad535386e
<pre>
Continue porting of graphics related types to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=251148">https://bugs.webkit.org/show_bug.cgi?id=251148</a>
rdar://104643561

Reviewed by Alex Christensen.

Continue the porting work of platform/graphics related types to the new
serialization format. This change includes the porting of:
    - CompositeOperator
    - BlendMode
    - GradientSpreadMethod
    - InterpolationQuality
    - LineCap
    - LineJoin
    - StrokeStyle
    - TextDrawingMode
    - DecodingMode
    - CompositeMode
    - DocumentMarkerLineStyleMode
    - ShadowRadiusMode
    - DropShadow
    - GenericCueData::Alignment
    - GenericCueData::Status
    - GenericCueData
    - ImageDecoderFrameInfo
    - DistantLightSource

* Source/WebCore/Headers.cmake:
* Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
(WebCore::Display::BorderPainter::drawBoxSideFromPath const):
(WebCore::Display::BorderPainter::drawLineForBoxSide const):
* Source/WebCore/platform/graphics/DecodingOptions.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::adjustLineToPixelBoundaries):
(WebCore::GraphicsContext::dashedLineCornerWidthForStrokeWidth const):
(WebCore::GraphicsContext::dashedLinePatternWidthForStrokeWidth const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextState.h:
* Source/WebCore/platform/graphics/GraphicsTypes.h:
(WebCore::DropShadow::hasOutsets const):
(WebCore::CompositeMode::encode const): Deleted.
(WebCore::CompositeMode::decode): Deleted.
(WebCore::DocumentMarkerLineStyle::encode const): Deleted.
(WebCore::DocumentMarkerLineStyle::decode): Deleted.
(WebCore::DropShadow::encode const): Deleted.
(WebCore::DropShadow::decode): Deleted.
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::FrameInfo::encode const): Deleted.
(WebCore::ImageDecoder::FrameInfo::decode): Deleted.
* Source/WebCore/platform/graphics/InbandGenericCue.h:
(WebCore::GenericCueData::decode): Deleted.
(WebCore::GenericCueData::encode const): Deleted.
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::dashedLineCornerWidthForStrokeWidth):
(WebCore::Cairo::dashedLinePatternWidthForStrokeWidth):
(WebCore::Cairo::State::setStrokeStyle):
(WebCore::Cairo::drawRect):
(WebCore::Cairo::drawLine):
(WebCore::Cairo::drawDotsForDocumentMarker):
(WebCore::Cairo::drawEllipse):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::drawLine):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawRect):
(WebCore::GraphicsContextCG::drawLine):
(WebCore::calculateDrawingMode):
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::colorForMarkerLineStyle):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawDotsForDocumentMarker::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawDotsForDocumentMarker::encode const):
(WebCore::DisplayList::DrawDotsForDocumentMarker::decode):
(WebCore::DisplayList::DrawDotsForDocumentMarker::isValid const): Deleted.
* Source/WebCore/platform/graphics/filters/DistantLightSource.h:
(WebCore::DistantLightSource::encode const): Deleted.
(WebCore::DistantLightSource::decode): Deleted.
* Source/WebCore/platform/graphics/filters/FELighting.h:
(WebCore::FELighting::encode const):
(WebCore::FELighting::decode):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.cpp:
(Nicosia::CairoOperationRecorder::drawLine):
* Source/WebCore/platform/graphics/win/GraphicsContextCGWin.cpp:
(WebCore::GraphicsContextCG::drawDotsForDocumentMarker):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline):
(WebCore::BorderPainter::drawBoxSideFromPath):
(WebCore::BorderPainter::drawLineForBoxSide):
* Source/WebCore/rendering/RenderDetailsMarker.cpp:
(WebCore::RenderDetailsMarker::paint):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIncompleteImageOutline const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::paint):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintProgressBar):
(WebCore::RenderThemeIOS::paintCheckbox):
(WebCore::RenderThemeIOS::paintRadio):
(WebCore::RenderThemeIOS::paintColorWellDecorations):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::paintAttachmentPlaceholderBorder):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarker):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::textDecorationStyleToStrokeStyle):
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
(WebCore::RenderMathMLBlock::paint):
* Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
(WebCore::RenderMathMLFraction::paint):
* Source/WebCore/rendering/mathml/RenderMathMLMenclose.cpp:
(WebCore::RenderMathMLMenclose::paint):
* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::applyStrokeStyleToContext):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/cg/BifurcatedGraphicsContextTestsCG.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/259648@main">https://commits.webkit.org/259648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56444560ae830aa1f0afdfe5e53465871e2390db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14429 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38234 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114613 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174802 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5362 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97664 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114512 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95055 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39556 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93939 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81233 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7751 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28049 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7864 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4725 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13895 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47593 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6658 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9653 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/graphics/ImageOrientation.h:
```diff
@@ -157,23 +172,17 @@ struct ImageOrientation {
-    WebCore::ImageOrientation::Orientation::OriginRightBottom,
-    WebCore::ImageOrientation::Orientation::OriginLeftBottom
-    >;
+enum class SerializableOrientation : int {
```
This isn't great.  Why not just make enum Orientation a nested enum class?

gavin-apple replied:
Proposing https://github.com/WebKit/WebKit/pull/9236 instead - will pull it out of this change.

achristensen07 commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
```diff
@@ -981,6 +982,35 @@ class DrawDotsForDocumentMarker {
 };
 
+template<class Encoder>
+void DrawDotsForDocumentMarker::encode(Encoder& encoder) const
```
Is this just work that still needs to be done?  It's not obvious why this is being added.

gavin-apple replied:
Yes, this needs to be ported but I didn't want to bloat this PR too much. Was going to put together a big DisplayList change at some point.

achristensen07 reviewed (approved):
Probably needs rebasing, double check fast/images/avif-as-image.html which seems like it might be related.

gavin-apple committed:
Continue porting of graphics related types to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=251148
rdar://104643561

Reviewed by Alex Christensen.

Continue the porting work of platform/graphics related types to the new
serialization format. This change includes the porting of:
    - CompositeOperator
    - BlendMode
    - GradientSpreadMethod
    - InterpolationQuality
    - LineCap
    - LineJoin
    - StrokeStyle
    - TextDrawingMode
    - DecodingMode
    - CompositeMode
    - DocumentMarkerLineStyleMode
    - ShadowRadiusMode
    - DropShadow
    - GenericCueData::Alignment
    - GenericCueData::Status
    - GenericCueData
    - ImageDecoderFrameInfo
    - DistantLightSource

* Source/WebCore/Headers.cmake:
* Source/WebCore/display/css/DisplayBoxDecorationPainter.cpp:
(WebCore::Display::BorderPainter::drawBoxSideFromPath const):
(WebCore::Display::BorderPainter::drawLineForBoxSide const):
* Source/WebCore/platform/graphics/DecodingOptions.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::adjustLineToPixelBoundaries):
(WebCore::GraphicsContext::dashedLineCornerWidthForStrokeWidth const):
(WebCore::GraphicsContext::dashedLinePatternWidthForStrokeWidth const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContextState.h:
* Source/WebCore/platform/graphics/GraphicsTypes.h:
(WebCore::DropShadow::hasOutsets const):
(WebCore::CompositeMode::encode const): Deleted.
(WebCore::CompositeMode::decode): Deleted.
(WebCore::DocumentMarkerLineStyle::encode const): Deleted.
(WebCore::DocumentMarkerLineStyle::decode): Deleted.
(WebCore::DropShadow::encode const): Deleted.
(WebCore::DropShadow::decode): Deleted.
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::FrameInfo::encode const): Deleted.
(WebCore::ImageDecoder::FrameInfo::decode): Deleted.
* Source/WebCore/platform/graphics/InbandGenericCue.h:
(WebCore::GenericCueData::decode): Deleted.
(WebCore::GenericCueData::encode const): Deleted.
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::dashedLineCornerWidthForStrokeWidth):
(WebCore::Cairo::dashedLinePatternWidthForStrokeWidth):
(WebCore::Cairo::State::setStrokeStyle):
(WebCore::Cairo::drawRect):
(WebCore::Cairo::drawLine):
(WebCore::Cairo::drawDotsForDocumentMarker):
(WebCore::Cairo::drawEllipse):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::drawLine):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawRect):
(WebCore::GraphicsContextCG::drawLine):
(WebCore::calculateDrawingMode):
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::colorForMarkerLineStyle):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawDotsForDocumentMarker::apply const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawDotsForDocumentMarker::encode const):
(WebCore::DisplayList::DrawDotsForDocumentMarker::decode):
(WebCore::DisplayList::DrawDotsForDocumentMarker::isValid const): Deleted.
* Source/WebCore/platform/graphics/filters/DistantLightSource.h:
(WebCore::DistantLightSource::encode const): Deleted.
(WebCore::DistantLightSource::decode): Deleted.
* Source/WebCore/platform/graphics/filters/FELighting.h:
(WebCore::FELighting::encode const):
(WebCore::FELighting::decode):
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm:
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.cpp:
(Nicosia::CairoOperationRecorder::drawLine):
* Source/WebCore/platform/graphics/win/GraphicsContextCGWin.cpp:
(WebCore::GraphicsContextCG::drawDotsForDocumentMarker):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline):
(WebCore::BorderPainter::drawBoxSideFromPath):
(WebCore::BorderPainter::drawLineForBoxSide):
* Source/WebCore/rendering/RenderDetailsMarker.cpp:
(WebCore::RenderDetailsMarker::paint):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIncompleteImageOutline const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::paint):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintProgressBar):
(WebCore::RenderThemeIOS::paintCheckbox):
(WebCore::RenderThemeIOS::paintRadio):
(WebCore::RenderThemeIOS::paintColorWellDecorations):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::paintAttachmentPlaceholderBorder):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarker):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::textDecorationStyleToStrokeStyle):
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
(WebCore::RenderMathMLBlock::paint):
* Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
(WebCore::RenderMathMLFraction::paint):
* Source/WebCore/rendering/mathml/RenderMathMLMenclose.cpp:
(WebCore::RenderMathMLMenclose::paint):
* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::applyStrokeStyleToContext):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/cg/BifurcatedGraphicsContextTestsCG.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/259648@main
